src/combinatorics/catalan.lean

```diff
@@ -129,3 +144,48 @@ by norm_num [catalan_eq_central_binom_div, nat.central_binom, nat.choose]
 
 lemma catalan_three : catalan 3 = 5 :=
 by norm_num [catalan_eq_central_binom_div, nat.central_binom, nat.choose]
+
+namespace tree
+open_locale tree
+
+/-- Given two finsets, find all trees that can be formed with
+  left child in `a` and right child in `b` -/
+@[simp] def pairwise_node (a : finset (tree unit)) (b : finset (tree unit)) : finset (tree unit) :=
+  (a ×ˢ b).map ⟨λ x, x.1 △ x.2, λ ⟨x₁, x₂⟩ ⟨y₁, y₂⟩, by { simp, tauto, }⟩
+
+/-- A finset of all trees with `n` nodes. See `mem_trees_of_nodes_eq` -/
+def trees_of_num_nodes_eq : ℕ → finset (tree unit)
+| 0 := {nil}
+| (n+1) := (finset.nat.antidiagonal n).attach.bUnion $ λ ijh,
+  have _ := nat.lt_succ_of_le (fst_le ijh.2),
+  have _ := nat.lt_succ_of_le (snd_le ijh.2),
+  pairwise_node (trees_of_num_nodes_eq ijh.1.1) (trees_of_num_nodes_eq ijh.1.2)
+
+@[simp] lemma trees_of_nodes_eq_zero : trees_of_num_nodes_eq 0 = {nil} :=
+by rw [trees_of_num_nodes_eq]
+
+lemma trees_of_nodes_eq_succ (n : ℕ) : trees_of_num_nodes_eq (n + 1) =
+  (nat.antidiagonal n).bUnion (λ ij, pairwise_node (trees_of_num_nodes_eq ij.1)
+    (trees_of_num_nodes_eq ij.2)) :=
+by { rw trees_of_num_nodes_eq, ext, simp [-pairwise_node], }
+
+@[simp] theorem mem_trees_of_nodes_eq {x : tree unit} {n : ℕ} :
+  x ∈ trees_of_num_nodes_eq n ↔ x.num_nodes = n :=
```

Comment:
Can you prove `mem_trees_of_nodes_eq_num_nodes : x ∈ trees_of_num_nodes_eq x.num_nodes`?

Reply:
@YaelDillies That follows from `mem_trees_of_nodes_eq` by `simp`. Do we want that as a separate lemma?

(`x ∈ trees_of_num_nodes_eq x.num_nodes` ==> `x.num_nodes = x.num_nodes` ==> `true`)

Reply:
Yes that looks like a good lemma for term mode proofs.